README.md
# Passphrase Generator

Creates a random passphrase using books from [Project Gutenberg](https://gutenberg.org).

## Description

*[Another](https://github.com/search?q=XKCD+Password+Generator)* passphrase generator based on [XKCD #936](https://xkcd.com/936). This command line generator, written in Python 3, creates a random passphrase primarily using the Python secrets module and specific books from [Project Gutenberg](https://gutenberg.org), and further generates information on an estimated time to crack the passphrase using [zxcvbn-python](https://github.com/dwolfhub/zxcvbn-python).

Default books used include: 
- [Alice's Adventures in Wonderland by Lewis Carroll](https://gutenberg.org/ebooks/11)
- [Frankenstein; Or, The Modern Prometheus by Mary Wollstonecraft Shelley](https://gutenberg.org/ebooks/84)
- [Winnie-the-Pooh by A. A. Milne](https://www.gutenberg.org/ebooks/67098)
- [A Manual or an Easy Method of Managing Bees by John M. Weeks](https://www.gutenberg.org/ebooks/27065)

Use due diligence with any password generator you use.

## Visuals

![Image of Passphrase Generator running on Windows 10.](passphrase_generator_image.png)

## Getting Started

### Dependencies

- The standalone executable has been tested on Microsoft Windows 10.
- The source code has been tested on Microsoft Windows 10 running Python 3.10.6.
- Dependencies required to run the source code consist of [zxcvbn-python](https://github.com/dwolfhub/zxcvbn-python) 4.4.28.

### Installing

The standalone executable is compressed in a zip folder. Once you unzip the folder, you may run the program by running ```passphrase_generator.exe```. Additionally, you may replace the four default books used, which are located in the ```books``` folder. The book file names must not be changed.

### Executing program

To run the program, open the file ```passphrase_generator.exe```. When the program has opened you will see information about the program, a passphrase, and the estimated time to crack the passphrase. You will also be presented with a yes or no question where you type ```y``` or ```n``` if you would like another passphrase generated. 

## Changelog

See the [CHANGELOG.md](https://github.com/PyDevJohn/passphrase-generator/blob/main/CHANGELOG.md) file for the latest changelog.

## License

This project is licensed under the MIT License. See the [LICENSE.md](https://github.com/PyDevJohn/passphrase-generator/blob/main/LICENSE) file for further details.

## Acknowledgments

- [PyInstaller](https://github.com/pyinstaller/pyinstaller)
- [zxcvbn-python](https://github.com/dwolfhub/zxcvbn-python)

## Contributing

Contributions are welcomed and appreciated! Please use [GitHub Docs](https://docs.github.com/) for guides on how to use Git, GitHub, and how to contribute. The only two requirements before submitting your code are the [pytest](https://github.com/pytest-dev/pytest) ```tests.py``` file passes its tests, and your code is run through [Black](https://github.com/psf/black).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### (Upcoming changes)
- N/A

## [1.0.0] - 2022-08-10
### Added
#### (New features)
- *[Another](https://github.com/search?q=XKCD+Password+Generator)* passphrase generator based on [XKCD #936](https://xkcd.com/936). This command line generator, written in Python 3, creates a random passphrase primarily using the Python secrets module and specific books from [Project Gutenberg](https://gutenberg.org), and further generates information on an estimated time to crack the passphrase using [zxcvbn-python](https://github.com/dwolfhub/zxcvbn-python).

### Changed
#### (Changes in existing functionality)
- N/A

### Deprecated
#### (Soon-to-be removed features)
- N/A

### Removed
#### (Now removed features)
- N/A

### Fixed
#### (Bug fixes)
- N/A

### Security
#### (Vulnerabilities)
- N/A


tests.py
import passphrase_generator
import pytest


def test_empty_random_passphrase():
    """Tests the passphrase_generator.random_passphrase() function to verify
    a string is returned.
    """

    random_passphrase = passphrase_generator.random_passphrase()

    if random_passphrase != "":
        assert True
    else:
        assert False


def test_empty_passphrase_strength():
    """Tests the passphrase_generator.passphrase_strength() function to verify
    a string is returned.
    """

    passphrase_strength = passphrase_generator.passphrase_strength("Test")

    if passphrase_strength != "":
        assert True
    else:
        assert False


if __name__ == "__main__":
    test_empty_random_passphrase()
    test_empty_passphrase_strength()


passphrase_generator.py
import string
import secrets
import random

import zxcvbn


def random_passphrase():
    """
    Creates a random passphrase primarily using the Python secrets module and
    specific books from Project Gutenberg.
    """

    # Book filenames relative location.
    filename_1 = "books/book_1.txt"
    filename_2 = "books/book_2.txt"
    filename_3 = "books/book_3.txt"
    filename_4 = "books/book_4.txt"

    # Try to open the book filenames as a string, replace digits and
    # punctuation, make words lowercase, and split words into list.
    try:
        with open(filename_1, encoding="utf-8") as f:
            book = f.read()
            for character in string.digits:
                book = book.replace(character, "")
            for character in string.punctuation:
                book = book.replace(character, "")
            word_list_1 = book.lower().split()
        with open(filename_2, encoding="utf-8") as f:
            book = f.read()
            for character in string.digits:
                book = book.replace(character, "")
            for character in string.punctuation:
                book = book.replace(character, "")
            word_list_2 = book.lower().split()
        with open(filename_3, encoding="utf-8") as f:
            book = f.read()
            for character in string.digits:
                book = book.replace(character, "")
            for character in string.punctuation:
                book = book.replace(character, "")
            word_list_3 = book.lower().split()
        with open(filename_4, encoding="utf-8") as f:
            book = f.read()
            for character in string.digits:
                book = book.replace(character, "")
            for character in string.punctuation:
                book = book.replace(character, "")
            word_list_4 = book.lower().split()
    # If book filenames cannot be found, show an error.
    except FileNotFoundError:
        print(
            "Error: A book file cannot be found. You may need to add a book text file."
        )
    # If reading book filenames were successful, choose a random list of words
    # from the secrets module, shuffle the position of the words using
    # the random module, and print the passphrase.
    else:
        random_word_1 = secrets.choice(word_list_1)
        random_word_2 = secrets.choice(word_list_2)
        random_word_3 = secrets.choice(word_list_3)
        random_word_4 = secrets.choice(word_list_4)

        # It appears the secrets module does not have a shuffle ability like
        # the random module, which is why it's being used here, but AFTER the
        # secrets module has run.
        random_words = [random_word_1, random_word_2, random_word_3, random_word_4]
        random.shuffle(random_words)

        passphrase = f"{random_words[0]} {random_words[1]} {random_words[2]} "
        passphrase += f"{random_words[3]}"

        return passphrase


def passphrase_strength(passphrase):
    """
    Uses zxcvbn-python to determine passphrase strength, and can give more
    information if needed.
    """

    # Follows the same principles as https://bitwarden.com/password-strength/
    passphrase_results = zxcvbn.zxcvbn(passphrase)
    crack_result = f"{passphrase_results['crack_times_display']['offline_slow_hashing_1e4_per_second'].title()}"

    return crack_result


def display_results():
    """
    Calls on random_passphrase() and passphrase_strength() functions and
    then displays the results.
    """

    TITLE = "Passphrase Generator"
    VERSION = "1.0.0"
    WEBSITE = "https://github.com/PyDevJohn/passphrase-generator"
    LICENSE = "MIT License"
    BUILT_WITH = "\t- PyInstaller (https://github.com/pyinstaller/pyinstaller)"
    BUILT_WITH += "\n\t- zxcvbn-python (https://github.com/dwolfhub/zxcvbn-python)"
    BOOKS_USED = "\t- Alice's Adventures in Wonderland by Lewis Carroll "
    BOOKS_USED += "(https://gutenberg.org/ebooks/11) \n\t- Frankenstein; Or, "
    BOOKS_USED += "The Modern Prometheus by Mary Wollstonecraft Shelley "
    BOOKS_USED += "(https://gutenberg.org/ebooks/84) \n\t- Winnie-the-Pooh by "
    BOOKS_USED += "A. A. Milne (https://www.gutenberg.org/ebooks/67098) \n\t- "
    BOOKS_USED += "A Manual or an Easy Method of Managing Bees by John M. "
    BOOKS_USED += "Weeks (https://www.gutenberg.org/ebooks/27065)"

    print(f"{TITLE}")
    print(f"Version: {VERSION}")
    print(f"Website: {WEBSITE}")
    print(f"License: {LICENSE}")
    print(f"Built with: \n{BUILT_WITH}")
    print(f"Books used for passphrases: \n{BOOKS_USED}")

    while True:
        passphrase = random_passphrase()
        crack_result = passphrase_strength(passphrase)

        print(f"\nPassphrase: {passphrase}")
        print(f"Estimated time to crack passphrase: {crack_result}\n")

        regenerate = input("Would you like to generate another passphrase? " "(y/n) ")
        if regenerate.lower().strip() == "n":
            break


# Run the below if this file is the main file being opened.
if __name__ == "__main__":
    display_results()
